internal/storage/sbom_store.go

```diff
@@ -0,0 +1,525 @@
+	}
+
+	if err := s.Versioner().UpdateObject(obj, 1); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	bytes, err := json.Marshal(sbom)
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	query, args, err := sq.Insert("sboms").
+		Columns("name", "namespace", "object").
+		Values(name, namespace, bytes).
+		ToSql()
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	_, err = s.db.ExecContext(ctx, query, args...)
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	outSBOM, ok := out.(*v1alpha1.SBOM)
+	if !ok {
+		return storage.NewInvalidObjError(key, fmt.Sprintf("unexpected out object type: %T", out))
+	}
+
+	*outSBOM = *sbom
+
+	if err := s.broadcaster.Action(watch.Added, sbom); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	return nil
+}
+
+// Delete removes the specified key and returns the value that existed at that spot.
+// If key didn't exist, it will return NotFound storage error.
+// If 'cachedExistingObject' is non-nil, it can be used as a suggestion about the
+// current version of the object to avoid read operation from storage to get it.
+// However, the implementations have to retry in case suggestion is stale.
+func (s *sbomStore) Delete(
+	ctx context.Context, key string, out runtime.Object, preconditions *storage.Preconditions,
+	validateDeletion storage.ValidateObjectFunc, _ runtime.Object,
+) error {
+	name, namespace := extractNameAndNamespace(key)
+	if name == "" || namespace == "" {
+		return storage.NewInternalErrorf("invalid key: %s", key)
+	}
+
+	tx, err := s.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+	defer func() {
+		if err := tx.Rollback(); !errors.Is(err, sql.ErrTxDone) {
+			log.Printf("failed to rollback transaction: %v", err)
+		}
+	}()
+
+	query, args, err := sq.Delete("sboms").
+		Where(sq.Eq{"name": name, "namespace": namespace}).
+		Suffix("RETURNING *").
+		ToSql()
+	if err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	sbomRecord := &sbomSchema{}
+	if err := tx.GetContext(ctx, sbomRecord, query, args...); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return storage.NewKeyNotFoundError(key, 0)
+		}
+		return storage.NewInternalError(err.Error())
+	}
+
+	if err := json.Unmarshal(sbomRecord.Object, out); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	if err := preconditions.Check(key, out); err != nil {
+		return err
+	}
+
+	if err := validateDeletion(ctx, out); err != nil {
+		return err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	if err := s.broadcaster.Action(watch.Deleted, out); err != nil {
+		return storage.NewInternalError(err.Error())
+	}
+
+	return nil
+}
+
+// Watch begins watching the specified key. Events are decoded into API objects,
+// and any items selected by 'p' are sent down to returned watch.Interface.
```

Comment:
I think `p` is now gone from the implementation of this function. Is it `opts` maybe?

Reply:
This is the original document from the Kubernetes authors. I will fix this in our codebase and try to fix upstream too. (p was referring to SelectionPredicate)